lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -236,29 +236,39 @@ defmodule Mix.Compilers.Elixir do
     end
   end
 
-  defp dependent_modules(sources, all_modules, pending_modules) do
+  defp dependent_runtime_modules(sources, all_modules, pending_modules) do
     changed_modules =
-      for module(module: module) = entry <- all_modules,
-          entry not in pending_modules,
+      for module <- all_modules,
+          module not in pending_modules,
           into: %{},
           do: {module, true}
 
-    Enum.flat_map(pending_modules, fn module(sources: source_files, module: module) ->
-      depending? =
-        Enum.any?(source_files, fn file ->
-          source(
-            compile_references: compile_refs,
-            struct_references: struct_refs,
-            runtime_references: runtime_refs
-          ) = List.keyfind(sources, file, source(:source))
-
-          has_any_key?(changed_modules, compile_refs) or
-            has_any_key?(changed_modules, struct_refs) or
-            has_any_key?(changed_modules, runtime_refs)
-        end)
-
-      if depending?, do: [module], else: []
-    end)
+    fixpoint_runtime_modules(sources, changed_modules, %{}, pending_modules)
+  end
+
+  defp fixpoint_runtime_modules(sources, changed, dependent, not_dependent) do
+    {new_dependent, not_dependent} =
+      Enum.reduce(not_dependent, {dependent, []}, fn module, {new_dependent, not_dependent} ->
+        depending? =
+          Enum.any?(module(module, :sources), fn file ->
+            source(runtime_references: runtime_refs) =
+              List.keyfind(sources, file, source(:source))
+
+            Enum.any?(changed, fn {module(module: module), true} -> module in runtime_refs end)
```

Comment:
has_any_key doesn't work here?

Reply:
No, the keys are `module()` records and we have a list of module atoms.

Reply:
Oh, I see. I think we can still store only the module name in the dependent map, no?

Reply:
Something like this (untested):

```elixir
  defp dependent_runtime_modules(sources, all_modules, pending_modules) do
    changed_modules =
      for module(module: module) = entry <- all_modules,
          entry not in pending_modules,
          into: %{},
          do: {module, true}

    fixpoint_runtime_modules(sources, changed_modules, %{}, pending_modules)
  end

  defp fixpoint_runtime_modules(sources, changed, dependent, not_dependent) do
    {new_dependent, not_dependent} =
      Enum.reduce(not_dependent, {dependent, []}, fn module, {new_dependent, not_dependent} ->
        depending? =
          Enum.any?(module(module, :source), fn file ->
            source(runtime_references: runtime_refs) =
              List.keyfind(sources, file, source(:source))

            has_any_key(changed, runtime_refs)
          end)

        if depending? do
          {Map.put(new_dependent, module(module, :module), true), not_dependent}
        else
          {new_dependent, [module | not_dependent]}
        end
      end)

    if map_size(dependent) != map_size(new_dependent) do
      fixpoint_runtime_modules(sources, new_dependent, new_dependent, not_dependent)
    else
      Map.keys(new_dependent)
    end
  end
```